ouroboros-consensus/test/consensus-test/Test/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -638,6 +836,65 @@ ppBlocks :: Point TestBlock -> [TestBlock] -> String
 ppBlocks a bs = ppPoint a <> " ] " <> intercalate " :> " (map ppBlock bs)
 
 ppTraceEvent :: TraceEvent -> String
-ppTraceEvent (Tick n, ev) = show n <> " | " <> case ev of
+ppTraceEvent (Tick n, m, ev) = show (n, m) <> " | " <> case ev of
     Left  cl -> "Client: "   <> show cl
     Right pt -> "Protocol: " <> show pt
+
+{-------------------------------------------------------------------------------
+  Classifying examples
+-------------------------------------------------------------------------------}
+
+data TickArrivalTimeStats a = OnlyNotEarly_SomeEarly {
+    onlyNotEarlyTATS   :: !a
+    -- ^ Logical ticks in which some headers are arriving but none are from the future.
+  , someEarlyTATS :: !a
+    -- ^ Logical ticks in which some headers are arriving from the future
+  }
+  deriving (Functor)
+  deriving (Generic)
+  deriving (Show) via (Quiet (TickArrivalTimeStats a))
+
+instance Monoid a => Monoid (TickArrivalTimeStats a) where
+    mempty = OnlyNotEarly_SomeEarly mempty mempty
+instance Semigroup a => Semigroup (TickArrivalTimeStats a) where
+    OnlyNotEarly_SomeEarly la lb <> OnlyNotEarly_SomeEarly ra rb =
+      OnlyNotEarly_SomeEarly (la <> ra) (lb <> rb)
+
+data ZOM = Zero | One | Many
+  deriving (Show)
+
+sizeZOM :: Set.Set a -> ZOM
+sizeZOM x = case Set.size x of
+    0 -> Zero
+    1 -> One
+    _ -> Many   -- NB negatives are impossible
+
+tickArrivalTimeStats :: [TraceEvent] -> TickArrivalTimeStats ZOM
+tickArrivalTimeStats events =
+    fmap sizeZOM $
+    OnlyNotEarly_SomeEarly {
+        onlyNotEarlyTATS = onlyNotEarly `Set.difference` someEarly
+      , someEarlyTATS    = someEarly
+      }
+  where
+    -- if you add a field to this pattern to avoid warnings, add it below too
+    OnlyNotEarly_SomeEarly _ _dummy = tickArrivalTimes events
+    OnlyNotEarly_SomeEarly {
+        onlyNotEarlyTATS = onlyNotEarly
+      , someEarlyTATS    = someEarly
+      } = tickArrivalTimes events
+
+-- | WARNING 'onlyNotEarlyTATS' is instead merely @pastTATS@ in this codomain
+tickArrivalTimes :: [TraceEvent] -> TickArrivalTimeStats (Set.Set Tick)
+tickArrivalTimes = foldMap $ \case
+    (n, now, Left (TraceDownloadedHeader hdr)) ->
+      let onset = RelativeTime $
+            scaleNominalDelay
+              (getSlotLength slotLength)
+              (unSlotNo (blockSlot hdr))
+          thisSlot = Set.singleton n
+      in
+      if now < onset
+      then OnlyNotEarly_SomeEarly { onlyNotEarlyTATS = Set.empty, someEarlyTATS = thisSlot }
+      else OnlyNotEarly_SomeEarly { onlyNotEarlyTATS = thisSlot, someEarlyTATS = Set.empty }
+    _ -> mempty
```

Comment:
This `now` here comes from `readTVar varLocalClock` which is only updated at the start of logical ticks, but it can be the case that during a logical tick in the ChainSync client, before the arrival of a header `H`, another header `H'` arrived from the future (and hence caused a call of `threadDelay`), such that `H` is *not* from the future when it arrives in the ChainSync client even though this logic would categorize it as such.

This probably doesn't make a huge difference, but it is sufficiently tricky that it maybe should at least be called out. Another option would be to use `getCurrentTime` in `pairWithNow`.

Reply:
Hmm. I think you're saying `H` and `H'` are both from the same logical tick, right?

In that case, it's fine. The semantics of `someEarlyTATs` is a count of how many logical ticks include at least one future header. It's not counting how many headers themselves.

Is that right?

(I also don't see how `getCurrentTime` would help.)

Reply:
> Hmm. I think you're saying `H` and `H'` are both from the same logical tick, right?
> 
> In that case, it's fine. The semantics of `someEarlyTATs` is a count of how many logical ticks include at least one future header. It's not counting how many headers themselves.
> 
> Is that right?

Ah right, that is a very good point, it is irrelevant whether just `H` or both `H` and `H` are considered to be from the future if they are from the same tick :+1:

Also, my comment doesn't apply anymore anyway as `now` is now backed by `getCurrentTime` via `SystemTime`, so it incorporates the `threadDelay`s in the ChainSync client.